backend/cfg/cfg_irc.ml

```diff
@@ -444,26 +447,80 @@ let rewrite : State.t -> Cfg_with_layout.t -> Reg.t list -> reset:bool -> unit =
         rewrite_body_and_terminator acc tl terminator)
   in
   Cfg.iter_blocks (Cfg_with_layout.cfg cfg_with_layout) ~f:(fun label block ->
-      (* CR xclerc for xclerc: we currently assume that a terminator does not
-         "define" a register that may be spilled. Calls are reasonably fine
-         since their result is in a precolored register. *)
-      assert (not (array_contains_spilled block.terminator.res));
       let body_needs_rewrite =
         instruction_list_contains_spilled block.body
         || array_contains_spilled block.terminator.arg
+        || array_contains_spilled block.terminator.res
       in
+      let trailing_spills = ref [] in
       if body_needs_rewrite
       then (
         if irc_debug
         then (
           log ~indent:2 "body of #%d, before:" label;
           log_body_and_terminator ~indent:3 block.body block.terminator);
-        block.body <- rewrite_body_and_terminator [] block.body block.terminator;
+        let new_body, spills =
+          rewrite_body_and_terminator [] block.body block.terminator
+        in
+        block.body <- new_body;
+        assert (List.length !trailing_spills = 0);
+        trailing_spills := spills;
         if irc_debug
         then (
           log ~indent:2 "and after:";
           log_body_and_terminator ~indent:3 block.body block.terminator;
-          log ~indent:2 "end")));
+          log ~indent:2 "end"));
+      if List.length !trailing_spills > 0
+      then (
+        let cfg = Cfg_with_layout.cfg cfg_with_layout in
+        let terminator = block.terminator in
+        (* Only terminators that are a raising operation can have spilled
+           results. *)
+        let next_label =
+          match terminator.desc with
+          | RaisingOp { op = _; label_after } -> label_after
+          | _ ->
+            Misc.fatal_errorf
+              "A terminator with spilled result that isn't a raising operation"
+        in
+        let next_block = Cfg.get_block_exn cfg next_label in
+        let spills = !trailing_spills in
+        assert (List.length spills > 0);
+        (* Make new block for the spills. *)
+        let spill_label = Cmm.new_label () in
+        let spill_block : Cfg.basic_block =
+          { start = spill_label;
+            body = spills;
+            terminator =
+              { desc = Cfg.Always next_label;
+                arg = [||];
+                res = [||];
+                dbg = terminator.dbg;
+                fdo = terminator.fdo;
+                live = terminator.live;
+                stack_offset = next_block.stack_offset;
```

Comment:
Should this be stack_offset of `block`, not of `next_block`?

Reply:
actually, it should be `terminator.stack_offset`

Reply:
Changed `spill_block.terminator.stack_offset` to `last_spill.stack_offset`.
Changed `spill_block.stack_offset` to `predecessor_block.terminator.stack_offset`.